mysterious-mice/spacebook/posts/views.py

```diff
@@ -0,0 +1,94 @@
+    GET parameter, then the ImagePosts returned will be limited to that user. This parameter will
+    be passed automatically if this page is navigated to from a rover's profile.
+    """
+
+    model = ImagePost
+    template_name = "posts/home.html"
+    context_object_name = "posts"
+    ordering = ["-date"]
+    paginate_by = 16
+
+    # if a valid username provided, only return that rover's posts
+    def get_queryset(self):
+        username = self.request.GET.get("username")
+        try:
+            rover = Rover.objects.get(username=username)
+            imageposts = ImagePost.objects.filter(rover=rover).order_by("-date")
+        except ObjectDoesNotExist:
+            imageposts = ImagePost.objects.all().order_by("-date")
+        return imageposts
+
+    # pass additional context if only displaying posts from one rover
+    def get_context_data(self, **kwargs):
+        # get the default context to append to
+        context = super(ImagePostListView, self).get_context_data(**kwargs)
+        username = self.request.GET.get("username")
+        try:
+            # if viewing posts by a specific rover, pass additional context to
+            # add GET data to URL and to display username in header
+            rover = Rover.objects.get(username=username)
+            context["username_get"] = f"username={ rover.username }&"
+            context["username"] = username
+        except ObjectDoesNotExist:
+            context["username_get"] = ""
+        return context
+
+
+class ImagePostDetailView(DetailView):
+    """
+    Provides a detail view for an ImagePost. If a valid username is provided as a GET parameter,
+    the next and previous links will navigate through only that rover's ImagePosts. Otherwise they
+    will navigate through all the ImagePosts on the site. This parameter will be passed
+    automatically if this page is navigated to from a rover's profile.
+    """
+
+    model = ImagePost
+    template_name = "posts/imagepost.html"
+
+    # get URLs for the previous/next photos
+    def get_context_data(self, **kwargs):
+        # get the default context to append to
+        context = super(ImagePostDetailView, self).get_context_data(**kwargs)
+
+        username = self.request.GET.get("username")
+        try:
+            # if viewing image from a profile, only navigate to that rover's images
+            rover = Rover.objects.get(username=username)
+            imageposts = list(ImagePost.objects.filter(rover=rover).order_by("-date"))
+            username_get = f"?username={username}"
+            context["username"] = username
+        except ObjectDoesNotExist:
+            # if no valid username, navigate to next/prev image regardless of rover
+            imageposts = list(ImagePost.objects.all().order_by("-date"))
+            username_get = ""
+
+        # if only one image returned, no links required
+        prev_image_pk = next_image_pk = "#"
+        if len(imageposts) > 1:
```

Comment:
looks a bit messy with all the ifs and else's. refactor it out might make it easier on the eyes.